### ISSUE TITLE
varOr() discards crossover offspring

### ISSUE THREAD

**Issue by jmetrius** [DEAP/deap#450]
https://github.com/DEAP/deap/blob/454b4f65a9c944ea2c90b38a75d384cddf524220/deap/algorithms.py#L229-L245

In the crossover section of this code, is there any reason to discard the second result of the crossover operation (ind2) instead of appending it to the offspring list?
Depending on the crossover probability this gives quite a good speedup (36% in my case).

**Comment by piprrr** [comment]
On Thu, Jan 2, 2020 at 3:29 PM jmetrius <notifications@github.com> wrote:

>
> https://github.com/DEAP/deap/blob/454b4f65a9c944ea2c90b38a75d384cddf524220/deap/algorithms.py#L229-L245
>
> In the crossover section of this code, is there any reason to discard the
> second result of the crossover operation (ind2) instead of appending it to
> the offspring list?
> Depending on the crossover probability this gives quite a good speedup
> (36% in my case).
>
Discarding one of the offspring from applying crossover introduces a source
of genetic drift, and should be avoided.

Cheers!

Mark
-- 
mcoletti@gmail.com

**Comment by jmetrius** [comment]
Thanks for that additional insight, Mark.
I attached my patch for the issue as a pull request.

**Comment by fmder** [comment]
Since the order of the offspring is random, there should be no generic drift. Using all the offspring skews the probabilities passed as argument.

**Comment by piprrr** [comment]
On Fri, Jan 3, 2020 at 8:04 AM François-Michel De Rainville <
notifications@github.com> wrote:

> Since the order of the offspring is random, there should be no generic
> drift. Using all the offspring skews the probabilities passed as argument.
>
> The order of the offspring is irrelevant.  Rather, it's the unique set of
alleles in the population that is at issue here; arbitrarily dropping one
of the offspring after crossover risks removing specific alleles from the
population.
-- 
mcoletti@gmail.com

**Comment by jmetrius** [comment]
> Using all the offspring skews the probabilities passed as argument.

Good point. I overlooked that.
However, the probabilities can be adjusted to allow for insertion of both children without skewing the offspring list.
I asked at [Stackexchange](https://math.stackexchange.com/questions/3496184/correcting-probabilities-when-outcome-is-used-twice) because I was not really sure how to go about that. The probability adjustments proposed there are now included in the pull request and work quite nice.
I tested the code by checking how many offspring children were constructed by each method this time.
While doing that i noticed a possible division by zero if the crossover probability is 1 and included a check.
